juju/charm.py

```diff
@@ -82,19 +117,145 @@ def __init__(self, framework, key, metadata):
-                                     RelationChangedEvent)
-                self.on.define_event(f'{relation_name}_relation_departed',
-                                     RelationDepartedEvent)
-                self.on.define_event(f'{relation_name}_relation_broken',
-                                     RelationBrokenEvent)
-
-        for storage_name in metadata.get('storage', ()):
-            self.on.define_event(f'{storage_name}_storage_attached',
-                                 StorageAttachedEvent)
-            self.on.define_event(f'{storage_name}_storage_detaching',
-                                 StorageDetachingEvent)
+        for relation_name in self.metadata.relations:
+            self.on.define_event(f'{relation_name}_relation_joined', RelationJoinedEvent)
+            self.on.define_event(f'{relation_name}_relation_changed', RelationChangedEvent)
+            self.on.define_event(f'{relation_name}_relation_departed', RelationDepartedEvent)
+            self.on.define_event(f'{relation_name}_relation_broken', RelationBrokenEvent)
+
+        for storage_name in metadata.storage:
+            self.on.define_event(f'{storage_name}_storage_attached', StorageAttachedEvent)
+            self.on.define_event(f'{storage_name}_storage_detaching', StorageDetachingEvent)
+
+
+class CharmMetadata:
+    """Object containing the metadata for the charm.
+
+    Supported fields are:
+
+        * ``name``
+        * ``summary``
+        * ``description``
+        * ``maintainers``
+        * ``tags``
+        * ``terms``
+        * ``series``
+        * ``subordinate``
+        * ``min_juju_version``
+        * ``requires``
+        * ``provides``
+        * ``peers``
+        * ``relations``
+        * ``storage``
+        * ``resources``
+        * ``payloads``
+        * ``extra_bindings``
+
+    The ``maintainers``, ``tags``, ``terms``, ``series``, and ``extra_bindings``
+    attributes are all tuples of strings.  The ``requires``, ``provides``,
+    `peers``, ``relations``, ``storage``, ``resources``, and ``payloads``
+    attributes are all mappings of names to instances of the respective
+    :class:`RelationDefinition`, :class:`StorageDefinition`, :class:`ResourceDefinition`,
+    or :class:`PayloadDefinition`.
+
+    The ``relations`` attribute is a convenience accessor which includes all of
+    the ``requires``, ``provides``, and ``peers`` :class:`RelationDefinition`_
+    items.  If needed, the role of the relation definition can be obtained from
+    its ``role`` attribute.
+    """
+    def __init__(self, raw=None):
+        raw = raw or {}
+        self.name = raw.get('name', '')
+        self.summary = raw.get('summary', '')
+        self.description = raw.get('description', '')
+        maintainers = []
+        if 'maintainer' in raw:
+            maintainers.append(raw['maintainer'])
+        if 'maintainers' in raw:
+            maintainers.extend(raw['maintainers'])
+        self.maintainers = tuple(maintainers)
```

Comment:
These seem to be more appropriate as lists.

Reply:
I went with tuples as they shouldn't be modified.  *shrug*